gobblin-tunnel/src/main/java/gobblin/tunnel/ReadWriteHandler.java

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (C) 2014-2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.tunnel;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.SelectionKey;
+import java.nio.channels.Selector;
+import java.nio.channels.SocketChannel;
+import java.util.concurrent.Callable;
+
+import static java.nio.channels.SelectionKey.OP_READ;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class handles the relaying of data back and forth between the Client-to-Tunnel and Tunnel-to-Proxy
+ * socket connections. This class is not thread safe.
+ */
+class ReadWriteHandler implements Callable<HandlerState> {
+  static final Logger LOG = LoggerFactory.getLogger(Tunnel.class);
+  private final SocketChannel proxy;
+  private final SocketChannel client;
+  private final Selector selector;
+  private final ByteBuffer buffer = ByteBuffer.allocate(1000000);
+  private HandlerState state = HandlerState.READING;
+
+  ReadWriteHandler(SocketChannel proxy, ByteBuffer mixedServerResponseBuffer, SocketChannel client, Selector selector)
+      throws IOException {
+    this.proxy = proxy;
+    this.client = client;
+    this.selector = selector;
+
+    // drain response that is not part of proxy's 200 OK and is part of data pushed from server, and push to client
+    if (mixedServerResponseBuffer.limit() > mixedServerResponseBuffer.position()) {
+      this.client.configureBlocking(true);
+      OutputStream clientOut = this.client.socket().getOutputStream();
+      clientOut.write(mixedServerResponseBuffer.array(), mixedServerResponseBuffer.position(),
+          mixedServerResponseBuffer.limit() - mixedServerResponseBuffer.position());
+      clientOut.flush();
```

Comment:
We don't want to close the client yet, since it will be used from this point forward to relay bytes back-and-forth. At this point we are just draining any part of the server-response that might have been intermingled with the proxy's 200 OK response before we start the main Tunnel operation.